Add Peer forwarder to doc website repo. (#1936)
opensearch-project/documentation-website: _data-prepper/peer-forwarder.md

---
layout: default
title: Peer Forwarder
nav_order: 12
---

> Peer Forwarder is an HTTP service which performs peer forwarding of `Event` between Data Prepper nodes for aggregation. Currently, supported by `aggregate`, `service_map_stateful`, `otel_trace_raw` processors.

Reviewer: Should this be “an Event” ?

> Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw` it's `traceId` by default and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).

Reviewer: ```suggestion
Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw`, it's `traceId` by default, and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).
```

> Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw` it's `traceId` by default and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).

Reviewer: We should avoid using “here” as the link text. Something around the page or section name.

> Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw` it's `traceId` by default and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).

Author: @ariamarble I think the identification key, but will check on this.

> Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw` it's `traceId` by default and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).

Author: @asifsmohammed what is configurable for aggregate?

> Peer Forwarder groups events based on the identification keys provided the processors. For `service_map_stateful` and `otel_trace_raw` it's `traceId` by default and can not be configured. It's configurable for `aggregate` processor using `identification_keys` configuration option. You can find more information about identification keys [here](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys).

Reviewer 2: Aggregate processor can be configured using `identification_keys` where user can specify the keys to use for Peer Forwarder. These keys can be anything from the user event data.

---

> Right now, peer discovery is provided by either a static list or by a DNS record lookup or AWS Cloudmap.

Reviewer: ```suggestion
Peer discovery is currently provided by either a static list or by a DNS record lookup or AWS Cloudmap.
```

# Discovery modes

> ## Static

Reviewer: Let’s make sure to place a line after each heading and before each code block. I’ve seen this break page rendering

> Static discover mode allows Data Prepper node to discover nodes using a list of IP addresses or domain names.

Reviewer: ```suggestion
Static discover mode allows a Data Prepper node to discover nodes using a list of IP addresses or domain names.
```

Your Data Prepper configuration needs to include:
* `aws_cloud_map_namespace_name` - Set to your Cloud Map Namespace name
* `aws_cloud_map_service_name` - Set to the service name within your specified Namespace
* `aws_region` - The AWS region where your namespace exists.
* `discovery_mode` - Set to `aws_cloud_map`

Your Data Prepper configuration can optionally include:
* `aws_cloud_map_query_parameters` - Key/value pairs to filter the results based on the custom attributes attached to an instance. Only instances that match all the specified key-value pairs are returned.

### Example configuration

```yaml
peer_forwarder:
  discovery_mode: aws_cloud_map
  aws_cloud_map_namespace_name: "my-namespace"
  aws_cloud_map_service_name: "data-prepper-cluster"
  aws_cloud_map_query_parameters:
    instance_type: "r5.xlarge"
  aws_region: "us-east-1"
```

### IAM policy with necessary permissions

> The Data Prepper must also be running with the necessary permissions. The following IAM policy shows the necessary permissions.

Reviewer: ```suggestion
The Data Prepper must also be running with the necessary permissions. The following IAM policy shows the necessary permissions:
```

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "CloudMapPeerForwarder",
            "Effect": "Allow",
            "Action": "servicediscovery:DiscoverInstances",
            "Resource": "*"
        }
    ]
}
```
---

> # Configuration

Reviewer: Should we be using more than 1 H1 heading on a single page?

## Optional configuration

> * `port`: An `int` between 0 and 65535 represents the port peer forwarder server is running on. Default value is `4994`.

Reviewer: I feel like this should be a table

* `aws_cloud_map_service_name`: A `String` representing the Cloud Map service when using AWS Cloud Map service discovery. Required if `discovery_mode` is set to `aws_cloud_map`.
* `aws_cloud_map_query_parameters`: A `Map` of Key/value pairs to filter the results based on the custom attributes attached to an instance. Only instances that match all the specified key-value pairs are returned.
* `buffer_size`: An `int` representing max number of unchecked records the buffer accepts (num of unchecked records = num of records written into the buffer + num of in-flight records not yet checked by the Checkpointing API). Default is `512`.
* `batch_size`: An `int` representing max number of records the buffer returns on read. Default is `48`.
* `aws_region`: A `String` represents the AWS region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`.
* `drain_timeout`: A `Duration` representing the wait time for the peer forwarder to complete processing data before shutdown.

> # SSL configuration

Reviewer: Same heading thoughts here. Maybe drop them all a level? Example responses etc are H4

> The SSL configuration for setting up trust manager for peer forwarding client to connect to other Data Prepper instances.

Reviewer: ```suggestion
The SSL configuration for setting up trust manager for the peer forwarding client to connect to other Data Prepper instances.
```

## Optional SSL configuration

> * `ssl`: A `boolean` that enables TLS/SSL. Default value is `true`.

Reviewer: This should also be a table

* `use_acm_certificate_for_ssl`: A `boolean` that enables TLS/SSL using certificate and private key from AWS Certificate Manager (ACM). Default is `false`.
* `acm_certificate_arn`: A `String` represents the ACM certificate ARN. ACM certificate take preference over S3 or local file system certificate. Required if `use_acm_certificate_for_ssl` is set to `true`.
* `acm_private_key_password`: A `String` that represents the ACM private key password which that will be used to decrypt the private key. If it's not provided, a random password will be generated.
* `acm_certificate_timeout_millis`: An `int` representing the timeout in milliseconds for ACM to get certificates. Default value is `120000`.
* `aws_region`: A `String` represents the AWS region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`.

```yaml
peer_forwarder:
  ssl: true
  ssl_certificate_file: "<cert-file-path>"
  ssl_key_file: "<private-key-file-path>"
```

# Authentication

> 

Reviewer: There should be a descriptive sentence after the heading.

* `authentication`(Optional) : A `Map` that enables mTLS. It can either be `mutual_tls` or `unauthenticated`. Default value is `unauthenticated`.
```yaml
peer_forwarder:
  authentication:
    mutual_tls:
```

# Metrics

> 

Reviewer: There should be a descriptive sentence after the heading.

> Core Peer Forwarder introduces the following custom metrics and all the metrics are prefixed by `core.peerForwarder`.

Reviewer: What are the custom metrics?

Author: @asifsmohammed Can we get some information on what the custom metrics are?

Reviewer 2: The metrics documented below are the ones. They're categorized into Timer, Counter and Gauge metrics.

I think the use of word `custom` here is confusing. These are the only metrics provided by Peer Forwarder.

# Timer

> 

Reviewer: There should be a descriptive sentence after the heading.

- `requestForwardingLatency`: measures latency of forwarding requests by peer forwarder client.
- `requestProcessingLatency`: measures latency of processing requests by peer forwarder server.

# Counter

- `requests`: measures total number of forwarded requests.
- `requestsFailed`: measures total number of failed requests. Requests with HTTP response code other than `200`.
- `requestsSuccessful`:  measures total number of successful requests. Requests with HTTP response code `200`.
- `requestsTooLarge`: measures total number of requests which are too large to be written to peer forwarder buffer. Requests with HTTP response code `413`.

> - `requestTimeouts`: measures total number of requests which timed out while writing content to peer forwarder buffer. Requests with HTTP response code `408`.

Reviewer: ```suggestion
- `requestTimeouts`: measures the total number of requests which timed out while writing content to peer forwarder buffer. Requests with HTTP response code `408`.
```

Reviewer: Same for the items below

- `requestsUnprocessable`: measures total number of requests which failed due to unprocessable entity. Requests with HTTP response code `422`.
- `badRequests`: measures total number of requests with bad request format. Requests with HTTP response code `400`.
- `recordsSuccessfullyForwarded`: measures total number of forwarded records successfully.
- `recordsFailedForwarding`: measures total number of records failed to be forwarded.
- `recordsToBeForwarded`: measures total number of records to be forwarded.
- `recordsToBeProcessedLocally`: measures total number of records to be processed locally.
- `recordsActuallyProcessedLocally`: measures total number of records actually processed locally. Sum of `recordsToBeProcessedLocally` and `recordsFailedForwarding`.
- `recordsReceivedFromPeers`: measures total number of records received from remote peers.

## Gauge

> 

Reviewer: There should be a descriptive sentence below the heading.

Peer Forwarder groups events based on the identification keys provided by the processors. For `service_map_stateful` and `otel_trace_raw`, the identification key is `traceId` by default and cannot be configured. The value is configurable for the `aggregate` processor using the `identification_keys` configuration option. You can find more information about [identification keys](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys) on the aggregate processor page.

Peer discovery is currently provided by either a static list or by a DNS record lookup or AWS Cloudmap.  

## Discovery modes

See the following information about discovery modes below.

### Static

Static discover mode allows a Data Prepper node to discover nodes using a list of IP addresses or domain names.
```yaml
peer_forwarder:4
  discovery_mode: static
  static_endpoints: ["data-prepper1", "data-prepper2"]
```

### DNS lookup

> We recommend using DNS discovery over static discovery when scaling out a Data Prepper cluster. The core concept is to configure a DNS provider to return a list of Data Prepper hosts when given a single domain name. This is a [DNS A Record](https://www.cloudflare.com/learning/dns/dns-records/dns-a-record/) which indicates a list of IP addresses of a given domain.

Reviewer 3: Should we avoid using the verbiage "We recommend" since this is an open source project? Suggestion: "DNS discovery is preferred over static..."

Add a comma after DNS A Record.

```yaml
peer_forwarder:
  discovery_mode: dns
  domain_name: "data-prepper-cluster.my-domain.net"
```

### AWS Cloud Map

[AWS Cloud Map](https://docs.aws.amazon.com/cloud-map/latest/dg/what-is-cloud-map.html) provides API-based service discovery as well as DNS-based service discovery.

> Peer forwarder can use the API-based service discovery. To support this you must have an existing namespace configured for API instance discovery. You can create a new one following the instructions provided by the [Cloud Map documentation](https://docs.aws.amazon.com/cloud-map/latest/dg/working-with-namespaces.html).

Reviewer 3: Is "Peer forwarder" a proper name? Should "forwarder" be capitalized?

Author: Peer Forwarder is a proper name, so it is capitalized.

Your Data Prepper configuration needs to include:
* `aws_cloud_map_namespace_name` - Set to your Cloud Map Namespace name
* `aws_cloud_map_service_name` - Set to the service name within your specified Namespace

> * `aws_region` - The AWS region where your namespace exists.

Reviewer 3: Capitalize "region."

Namespace is capitalized in the above bullets. Capitalize here as well?

* `discovery_mode` - Set to `aws_cloud_map`

Your Data Prepper configuration can optionally include:

> * `aws_cloud_map_query_parameters` - Key/value pairs to filter the results based on the custom attributes attached to an instance. Only instances that match all the specified key-value pairs are returned.

Reviewer 3: Should "Key/value pairs" be spelled "Key-value pairs?" See the spelling convention in the second sentence.

### Example configuration

```yaml
peer_forwarder:
  discovery_mode: aws_cloud_map
  aws_cloud_map_namespace_name: "my-namespace"
  aws_cloud_map_service_name: "data-prepper-cluster"
  aws_cloud_map_query_parameters:
    instance_type: "r5.xlarge"
  aws_region: "us-east-1"
```

### IAM policy with necessary permissions

The Data Prepper must also be running with the necessary permissions. The following IAM policy shows the necessary permissions.

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "CloudMapPeerForwarder",
            "Effect": "Allow",
            "Action": "servicediscovery:DiscoverInstances",
            "Resource": "*"
        }
    ]
}
```
---
## Configuration

See the following for optional configuration values.

### Optional configuration

| Value | Description |
| ----  | ----------- |

> | `port` | An `int` between 0 and 65535 represents the port peer forwarder server is running on. Default value is `4994`.|

Reviewer 3: Should "peer forwarder" be capitalized?

| `request_timeout` | Duration - An `int` representing the request timeout in milliseconds for Peer Forwarder HTTP server. Default value is `10000`. |

> | `server_thread_count` | An `int` representing number of threads used by Peer Forwarder server. Defaults to `200`.|

Reviewer 3: Should this read "used by **the** Peer Forwarder server?" If so, make global changes throughout doc as appropriate.

> | `client_thread_count` | An `int` representing number of threads used by Peer Forwarder client. Defaults to `200`.|

Reviewer 3: Should this read "used by **the** Peer Forwarder client?"

| `maxConnectionCount`  |  An `int` representing maximum number of open connections for Peer Forwarder server. Default value is `500`. |

> | `discovery_mode` | A `String` representing the peer discovery mode to be used. Allowable values are `local_node`, `static`, `dns`, and `aws_cloud_map`. Defaults to `local_node` which processes events locally.  |

Reviewer 3: Add comma after `local_node`.

| `static_endpoints` |  A `list` containing endpoints of all Data Prepper instances. Required if `discovery_mode` is set to `static`. |
|  `domain_name` | A `String` representing single domain name to query DNS against. Typically, used by creating multiple [DNS A Records](https://www.cloudflare.com/learning/dns/dns-records/dns-a-record/) for the same domain. Required if `discovery_mode` is set to `dns`.  |

> | `aws_cloud_map_namespace_name`  | A `String` representing the Cloud Map namespace when using AWS Cloud Map service discovery. Required if `discovery_mode` is set to `aws_cloud_map`.  |

Reviewer 3: Should "namespace" be capitalized? It's treated as a proper name in preceding sections.

| `aws_cloud_map_service_name` |  A `String` representing the Cloud Map service when using AWS Cloud Map service discovery. Required if `discovery_mode` is set to `aws_cloud_map`. |

> | `aws_cloud_map_query_parameters`  |  A `Map` of Key/value pairs to filter the results based on the custom attributes attached to an instance. Only instances that match all the specified key-value pairs are returned. |

Reviewer 3: Should "Key/value" pairs be spelled "key-value pairs?"

| `ssl_key_file`| A `String` represents the SSL key file path or AWS S3 path. S3 path example `s3://<bucketName>/<path>`. Defaults to `config/default_private_key.pem` which is default private key file. Read more about how the private key file is generated [here](https://github.com/opensearch-project/data-prepper/tree/main/examples/certificates). |
| `ssl_insecure_disable_verification` | A `boolean` that disables the verification of server's TLS certificate chain. Default value is `false`. |
| `ssl_fingerprint_verification_only` | A `boolean` that disables the verification of server's TLS certificate chain and instead verifies only the certificate fingerprint. Default value is `false`. |
| `use_acm_certificate_for_ssl` | A `boolean` that enables TLS/SSL using certificate and private key from AWS Certificate Manager (ACM). Default is `false`. |
| `acm_certificate_arn`| A `String` represents the ACM certificate ARN. ACM certificate take preference over S3 or local file system certificate. Required if `use_acm_certificate_for_ssl` is set to `true`. |

> | `acm_private_key_password` | A `String` that represents the ACM private key password which that will be used to decrypt the private key. If it's not provided, a random password will be generated. |

Reviewer 3: Add comma after "password" and delete "that." Or, delete "which."

| `acm_certificate_timeout_millis` | An `int` representing the timeout in milliseconds for ACM to get certificates. Default value is `120000`. |

> | `aws_region` | A `String` represents the AWS region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`. |

Reviewer 3: Capitalize "region" and add comma after "S3."

```yaml
peer_forwarder:
  ssl: true
  ssl_certificate_file: "<cert-file-path>"
  ssl_key_file: "<private-key-file-path>"
```

## Authentication

This section describes optional authentication.

`authentication`(optional) : A `Map` that enables mTLS. It can either be `mutual_tls` or `unauthenticated`. The default value is `unauthenticated`.

```yaml
peer_forwarder:
  authentication:
    mutual_tls:
```

## Metrics

This section describes custom metrics capabilities, including timer, counter, and gauge.

Core Peer Forwarder introduces the following custom metrics and all the metrics are prefixed by `core.peerForwarder`.

### Timer

This section describes the metrics feature, timer.

- `requestForwardingLatency`: measures latency of forwarding requests by peer forwarder client.
- `requestProcessingLatency`: measures latency of processing requests by peer forwarder server.

### Counter

See the table below for counter metric options.

> | Value | Description |

Reviewer 3: Global comment: Preceding tables initial cap the first word under Description. Should this table follow same?

| ----- | ----------- |
| `requests`| measures total number of forwarded requests. |
| `requestsFailed`| measures total number of failed requests. Requests with HTTP response code other than `200`. |
| `requestsSuccessful`|  measures total number of successful requests. Requests with HTTP response code `200`. |

> | `requestsTooLarge`| measures total number of requests which are too large to be written to peer forwarder buffer. Requests with HTTP response code `413`. |

Reviewer 3: Change "which" to "that." Should "peer forwarder" be capitalized?

> | `requestTimeouts`| measures the total number of requests which timed out while writing content to peer forwarder buffer. Requests with HTTP response code `408`. |

Reviewer 3: Change "which" to "that."

> | `requestsUnprocessable`| measures total number of requests which failed due to unprocessable entity. Requests with HTTP response code `422`. |

Reviewer 3: Change "which" to "that."

| `badRequests`| measures total number of requests with bad request format. Requests with HTTP response code `400`. |
| `recordsSuccessfullyForwarded`| measures total number of forwarded records successfully. |
| `recordsFailedForwarding`| measures total number of records failed to be forwarded. |
| `recordsToBeForwarded` | measures total number of records to be forwarded. |
| `recordsToBeProcessedLocally` | measures total number of records to be processed locally. |

> | `recordsActuallyProcessedLocally`| measures total number of records actually processed locally. Sum of  recordsToBeProcessedLocally` and `recordsFailedForwarding`. |

Reviewer 3: Delete extra space before `recordsToBeProcessedLocally`.

| `recordsReceivedFromPeers`| measures total number of records received from remote peers. |

### Gauge

This section describes gauge metric options.

> `peerEndpoints` measures number of dynamically discovered peer data-prepper endpoints. For `static` mode, the size is fixed.

Reviewer 3: Is "data-prepper"spelled this way intentionally? That is instead of "peer Data Prepper endpoints?"

---
layout: default
title: Peer Forwarder
nav_order: 12
---

# Peer Forwarder

Peer Forwarder is an HTTP service that performs peer forwarding of an `event` between Data Prepper nodes for aggregation. Currently, Peer Forwarder is supported by `aggregate`, `service_map_stateful`, `otel_trace_raw` processors.

> Peer Forwarder groups events based on the identification keys provided by the processors. For `service_map_stateful` and `otel_trace_raw`, the identification key is `traceId` by default and cannot be configured. The `aggregate` processor is configured using the `identification_keys` configuration option. From here, you can specify which keys to use for Peer Forwarder. You can find more information about [identification keys](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys) on the aggregate processor page.

Author: @natebower We can. Should we change how we link to the identification keys section and just add a link at the end of the section to the aggregate processor page as a whole?

> Peer Forwarder groups events based on the identification keys provided by the processors. For `service_map_stateful` and `otel_trace_raw`, the identification key is `traceId` by default and cannot be configured. The `aggregate` processor is configured using the `identification_keys` configuration option. From here, you can specify which keys to use for Peer Forwarder. You can find more information about [identification keys](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/aggregate-processor#identification_keys) on the aggregate processor page.

Reviewer 4: I would probably structure it as "See the [Aggregate Processor page] for more information about identification keys."

### DNS lookup
DNS discovery is preferred over static discovery when scaling out a Data Prepper cluster. The core concept is to configure a DNS provider to return a list of Data Prepper hosts when given a single domain name. This is a [DNS A Record](https://www.cloudflare.com/learning/dns/dns-records/dns-a-record/), which indicates a list of IP addresses of a given domain.

```yaml
peer_forwarder:
  discovery_mode: dns
  domain_name: "data-prepper-cluster.my-domain.net"
```

### AWS Cloud Map

[AWS Cloud Map](https://docs.aws.amazon.com/cloud-map/latest/dg/what-is-cloud-map.html) provides API-based service discovery as well as DNS-based service discovery.

Peer Forwarder can use the API-based service discovery. To support this, you must have an existing Namespace configured for API instance discovery. You can create a new one following the instructions provided by the [Cloud Map documentation](https://docs.aws.amazon.com/cloud-map/latest/dg/working-with-namespaces.html).

Your Data Prepper configuration needs to include:
* `aws_cloud_map_namespace_name` - Set to your Cloud Map Namespace name
* `aws_cloud_map_service_name` - Set to the service name within your specified Namespace

> * `aws_region` - The AWS Region where your Namespace exists.

Author: I think we should to remain consistent with previous lines in the same section.

* `aws_cloud_map_query_parameters` - Key-value pairs to filter the results based on the custom attributes attached to an instance. Only instances that match all the specified key-value pairs are returned.

#### Example configuration

```yaml
peer_forwarder:
  discovery_mode: aws_cloud_map
  aws_cloud_map_namespace_name: "my-namespace"
  aws_cloud_map_service_name: "data-prepper-cluster"
  aws_cloud_map_query_parameters:
    instance_type: "r5.xlarge"
  aws_region: "us-east-1"
```

### IAM policy with necessary permissions

The Data Prepper must also be running with the necessary permissions. The following IAM policy shows the necessary permissions:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "CloudMapPeerForwarder",
            "Effect": "Allow",
            "Action": "servicediscovery:DiscoverInstances",
            "Resource": "*"
        }
    ]
}
```
---
## Configuration

See the following for optional configuration values.

### Optional configuration

| Value | Description |
| ----  | ----------- |
| `port` | An `int` between 0 and 65535 represents the port Peer Forwarder server is running on. Default value is `4994`.|

> | `request_timeout` | Duration - An `int` representing the request timeout in milliseconds for Peer Forwarder HTTP server. Default value is `10000`. |

Author: I think if we move "duration" after "timeout", it makes more sense. ("...representing the request timeout duration in milliseconds...").

### Optional SSL configuration

See the table below for optional SSL configuration descriptions.

| Value | Description |
| ----- | ----------- |
| `ssl` | A `boolean` that enables TLS/SSL. Default value is `true`. |
| `ssl_certificate_file`| A `String` representing the SSL certificate chain file path or AWS S3 path. S3 path example `s3://<bucketName>/<path>`. Defaults to the default certificate file,`config/default_certificate.pem`. Read more about how the certificate file is generated at the [Default Certificates](https://github.com/opensearch-project/data-prepper/tree/main/examples/certificates) page. |
| `ssl_key_file`| A `String` represents the SSL key file path or AWS S3 path. S3 path example `s3://<bucketName>/<path>`. Defaults to `config/default_private_key.pem` which is default private key file. Read more about how the private key file is generated at the [Default Certificates](https://github.com/opensearch-project/data-prepper/tree/main/examples/certificates) page. |
| `ssl_insecure_disable_verification` | A `boolean` that disables the verification of server's TLS certificate chain. Default value is `false`. |

> | `ssl_fingerprint_verification_only` | A `boolean` that disables the verification of server's TLS certificate chain and instead verifies only the certificate fingerprint. Default value is `false`. |

Author: @natebower Should we change to "A 'Boolean' disabling..." to remain consistent with other items in this section? Right now the rest "An XYZ representing..." are all in gerund form.

> | `ssl_fingerprint_verification_only` | A `boolean` that disables the verification of server's TLS certificate chain and instead verifies only the certificate fingerprint. Default value is `false`. |

Reviewer 4: No, I would leave as is. It sounds like the Boolean takes the action once, whereas "representing" is continuous.

| `use_acm_certificate_for_ssl` | A `boolean` that enables TLS/SSL using certificate and private key from AWS Certificate Manager (ACM). Default is `false`. |
| `acm_certificate_arn`| A `String` represents the ACM certificate ARN. ACM certificate take preference over S3 or local file system certificate. Required if `use_acm_certificate_for_ssl` is set to `true`. |
| `acm_private_key_password` | A `String` that represents the ACM private key password that will be used to decrypt the private key. If it's not provided, a random password will be generated. |

> | `acm_certificate_timeout_millis` | An `int` representing the timeout in milliseconds for ACM to get certificates. Default value is `120000`. |

Author: @natebower Please clarify what specifically needs extra revision here.

> | `acm_certificate_timeout_millis` | An `int` representing the timeout in milliseconds for ACM to get certificates. Default value is `120000`. |

Reviewer 4: Something like "An `int` representing the timeout in milliseconds required for ACM to get certificates."

> | `aws_region` | A `String` represents the AWS Region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`. |

Author: @natebower Please clarify what specifically needs extra revision here. Right now, it reads: "  A `String` represents the AWS Region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file`,  and `ssl_key_file` is set to the `AWS S3` path, or if `discovery_mode` is set to `aws_cloud_map`."

> | `aws_region` | A `String` represents the AWS Region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`. |

Reviewer 4: As written, the sentences are difficult to parse because they are confusingly organized. They need to be simplified/clarified.

> | `aws_region` | A `String` represents the AWS Region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`. |

Author: @cwillum This comment will need some extra attention in doc review, FYI!

> | `aws_region` | A `String` represents the AWS Region to use `ACM`, `S3` or `AWS Cloud Map`. Required if `use_acm_certificate_for_ssl` is set to `true` or `ssl_certificate_file` and `ssl_key_file` is `AWS S3` path or if `discovery_mode` is set to `aws_cloud_map`. |

Author: @Naarcha-AWS Please take a look. Thanks!

#### Example config

```yaml
peer_forwarder:
  ssl: true
  ssl_certificate_file: "<cert-file-path>"
  ssl_key_file: "<private-key-file-path>"
```

## Authentication

This section describes optional authentication.

`authentication`(optional) : A `Map` that enables mTLS. It can either be `mutual_tls` or `unauthenticated`. The default value is `unauthenticated`.

```yaml
peer_forwarder:
  authentication:
    mutual_tls:
```

## Metrics

This section describes Peer Forwarder's metrics capabilities, including timer, counter, and gauge.

Core Peer Forwarder introduces the following custom metrics and all the metrics are prefixed by `core.peerForwarder`.

### Timer

This section describes the timer metrics feature.

- `requestForwardingLatency`: measures latency of forwarding requests by the Peer Forwarder client.
- `requestProcessingLatency`: measures latency of processing requests by Peer Forwarder server.

### Counter

See the table below for counter metric options.

| Value | Description |
| ----- | ----------- |
| `requests`| Measures total number of forwarded requests. |

> | `requestsFailed`| Measures total number of failed requests. Requests with HTTP response code other than `200`. |

Author: @chenqi0805 Can you clarify what the second sentence is attempting to communicate?

> | `requestsFailed`| Measures total number of failed requests. Requests with HTTP response code other than `200`. |

Author: @dlvenable Do you have any feedback on this?

> | `requestsSuccessful`|  Measures total number of successful requests. Requests with HTTP response code `200`. |

Author: @chenqi0805 Also need clarification on the second sentence here. It looks like some information may have been cut off.

| `requestsTooLarge`| Measures total number of requests that are too large to be written to Peer Forwarder buffer. Requests with HTTP response code `413`. |
| `requestTimeouts`| Measures the total number of requests that timed out while writing content to Peer Forwarder buffer. Requests with HTTP response code `408`. |
| `requestsUnprocessable`| Measures total number of requests that failed due to unprocessable entity. Requests with HTTP response code `422`. |

> | `badRequests`| Measures total number of requests with bad request format. Requests with HTTP response code `400`. |

Author: @chenqi0805 Can you please clarify what "with bad request format" means? Should we use "Measures the total number of requests that are in an incorrect request format." ?

> | `recordsSuccessfullyForwarded`| Measures total number of forwarded records successfully. |

Reviewer 4: ```suggestion
| `recordsSuccessfullyForwarded`| Measures the total number of successfully forwarded records. |
```

> | `recordsFailedForwarding`| Measures total number of records failed to be forwarded. |

Reviewer 4: ```suggestion
| `recordsFailedForwarding`| Measures the total number of records that fail to be forwarded. |
```

> | `recordsToBeForwarded` | Measures total number of records to be forwarded. |

Reviewer 4: ```suggestion
| `recordsToBeForwarded` | Measures the total number of records to be forwarded. |
```

> | `recordsToBeProcessedLocally` | Measures total number of records to be processed locally. |

Reviewer 4: ```suggestion
| `recordsToBeProcessedLocally` | Measures the total number of records to be processed locally. |
```

> | `recordsActuallyProcessedLocally`| Measures total number of records actually processed locally. Sum of `recordsToBeProcessedLocally` and `recordsFailedForwarding`. |

Reviewer 4: ```suggestion
| `recordsActuallyProcessedLocally`| Measures the total number of records actually processed locally. Sum of `recordsToBeProcessedLocally` and `recordsFailedForwarding`. |
```

Reviewer 4: Please clarify the second sentence. "This value is the sum of..."?

> | `recordsReceivedFromPeers`| Measures total number of records received from remote peers. |

Reviewer 4: ```suggestion
| `recordsReceivedFromPeers`| Measures the total number of records received from remote peers. |
```

### Gauge

> This section describes gauge metric options.

Reviewer 4: Global: We don't need this type of introductory sentence in every section. It's sufficient to just provide the relevant text. Please revise where applicable.

> `peerEndpoints` Measures number of dynamically discovered peer Data Prepper endpoints. For `static` mode, the size is fixed.

Reviewer 4: ```suggestion
`peerEndpoints` Measures the number of dynamically discovered peer Data Prepper endpoints. For `static` mode, the size is fixed.
```
